vendor/github.com/Juniper/contrail-go-api/keystone.go

```diff
@@ -135,6 +155,34 @@ func (kClient *KeystoneClient) Authenticate() error {
 	return nil
 }
 
+func (kClient *KeepaliveKeystoneClient) needsRefreshing() bool {
+	if kClient.current == nil {
+		return true
+	}
+
+	return time.Now().UTC().After(kClient.nextRefresh.UTC())
+}
+
+func (kClient *KeepaliveKeystoneClient) AddAuthentication(req *http.Request) error {
+	if kClient.needsRefreshing() {
+		err := kClient.Authenticate()
+		if err != nil {
+			return err
+		}
```

Comment:
I'm wondering if it would be better to set `kClient.current` to `nil` and then call `AddAuthentication` method from `KeystoneClient`, like this: https://play.golang.org/p/kDAanRyV9D
```
package main

import (
	"fmt"
)

type A struct {}
type B struct { A }

func (a* A) foo() {
	fmt.Println("greetings from A")
}

func (b* B) foo() {
	fmt.Println("greetings from B")
	b.A.foo()
}

func main() {
	fmt.Println("Hello, playground")
	var myB B
	myB.foo()
}
```
The benefit is that we avoid some code duplication: calling `Authenticate`, error handling and setting the `X-Auth-Token` header.